### ISSUE TITLE
Improve CLI

### ISSUE THREAD

**Issue by andylokandy** [elba/elba#12]
I've collected logs of some of my familiar tools. I think they help refining our log style.

pip
```
pip install numpy
Collecting numpy                                                  
  Downloading https://files.pythonhosted.org/packages/53/d1/2499794.whl (13.5MB)
    100% |████████████████████████████████| 13.5MB 87kB/s         
Installing collected packages: numpy
Successfully installed numpy-1.15.0
You are using pip version 9.0.3, however version 18.0 is available.
You should consider upgrading via the 'python -m pip install --upgrade pip' command.
```

cargo
```
cargo install cargo-tree
    Updating registry `https://github.com/rust-lang/crates.io-index`
 Downloading cargo-tree v0.19.0
  Installing cargo-tree v0.19.0
 Downloading cargo v0.28.0
 Downloading ignore v0.4.3
 Downloading quote v0.6.6
 Downloading syn v0.14.8
 Downloading globset v0.4.1
   Compiling proc-macro2 v0.4.11
   Building [===================================================>   ] 124/130: libgit2-sys

   ...
   Compiling vcpkg v0.2.4
   Compiling winapi v0.3.5
   Compiling void v1.0.2
   Compiling matches v0.1.7
   Compiling cargo-tree v0.19.0
    Finished release [optimized] target(s) in 4m 59s
  Installing C:\Users\Andy\.cargo\bin\cargo-tree.exe
```

npm
```
npm install -g parcel-bundler
npm WARN deprecated browserslist@1.7.7: Browserslist 2 could fail on reading Browserslist >3.0 config used in other tools.
[..................] / fetchMetadata: sill resolveWithNewModule parcel-bundler@1.9.7 checking installable status
C:\Users\Andy\AppData\Roaming\npm\parcel -> C:\Users\Andy\AppData\Roaming\npm\node_modules\parcel-bundler\bin\cli.js

> deasync@0.1.13 install C:\Users\Andy\AppData\Roaming\npm\node_modules\parcel-bundler\node_modules\deasync
> node ./build.js

`win32-x64-node-8` exists; testing
Binary is fine; exiting

> parcel-bundler@1.9.7 postinstall C:\Users\Andy\AppData\Roaming\npm\node_modules\parcel-bundler
> node -e "console.log('\u001b[35m\u001b[1mLove Parcel? You can now donate to our open collective:\u001b[22m\u001b[39m\n > \u001b[34mhttps://opencollective.com/parcel/donate\u001b[0m')"

Love Parcel? You can now donate to our open collective:
 > https://opencollective.com/parcel/donate
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@1.2.4 (node_modules\parcel-bundler\node_modules\fsevents):
npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@1.2.4: wanted {"os":"darwin","arch":"any"} (current: {"os":"win32","arch":"x64"})

+ parcel-bundler@1.9.7
added 681 packages in 75.745s
```

purescript(pulp)
```
pulp run
Building project in /Users/andy/hello_world
Compiling Record.Unsafe
Compiling Type.Data.RowList
Compiling Type.Data.Row
Compiling Data.Symbol
Compiling Data.Boolean
Compiling Control.Semigroupoid
Compiling Control.Category
Compiling Data.Show
Compiling Data.Void
Compiling Data.Unit
Compiling Data.Semiring
Compiling Data.Semigroup
Compiling Data.Ring
Compiling Data.Eq
Compiling Main
Build successful.
Hello world!
```

parcel
```
parcel index.html
Server running at http://localhost:1234
/ Building hmr-runtime.js...

...
√  Built in 1.37s.
```

sbt
```
sbt
Project does not exist, create new project? (y/N/s) y
Name: sample
Organization: com.twitter
Version [1.0]: 1.0-SNAPSHOT
Scala version [2.7.7]: 2.8.1
sbt version [0.7.4]:      
Getting Scala 2.7.7 ...
:: retrieving :: org.scala-tools.sbt#boot-scala
	confs: [default]
	2 artifacts copied, 0 already retrieved (9911kB/221ms)
Getting org.scala-tools.sbt sbt_2.7.7 0.7.4 ...
:: retrieving :: org.scala-tools.sbt#boot-app
	confs: [default]
	15 artifacts copied, 0 already retrieved (4096kB/167ms)
[success] Successfully initialized directory structure.
Getting Scala 2.8.1 ...
:: retrieving :: org.scala-tools.sbt#boot-scala
	confs: [default]
	2 artifacts copied, 0 already retrieved (15118kB/386ms)
[info] Building project sample 1.0-SNAPSHOT against Scala 2.8.1
[info]    using sbt.DefaultProject with sbt 0.7.4 and Scala 2.7.7

> test
[info] 
[info] == compile ==
[info]   Source analysis: 0 new/modified, 0 indirectly invalidated, 0 removed.
[info] Compiling main sources...
[info] Nothing to compile.
[info]   Post-analysis: 3 classes.
[info] == compile ==
[info] 
[info] == test-compile ==
[info]   Source analysis: 0 new/modified, 0 indirectly invalidated, 0 removed.
[info] Compiling test sources...
[info] Nothing to compile.
[info]   Post-analysis: 10 classes.
[info] == test-compile ==
[info] 
[info] == copy-test-resources ==
[info] == copy-test-resources ==
[info] 
[info] == copy-resources ==
[info] == copy-resources ==
[info] 
[info] == test-start ==
[info] == test-start ==
[info] 
[info] == com.twitter.sample.SimpleParserSpec ==
[info] SimpleParserSpec
[info] SimpleParser should
[info]   + work with basic tweet
[info] == com.twitter.sample.SimpleParserSpec ==
[info] 
[info] == test-complete ==
[info] == test-complete ==
[info] 
[info] == test-finish ==
[info] Passed: : Total 1, Failed 0, Errors 0, Passed 1, Skipped 0
[info]  
[info] All tests PASSED.
[info] == test-finish ==
[info] 
[info] == test-cleanup ==
[info] == test-cleanup ==
[info] 
[info] == test ==
[info] == test ==
[success] Successful.
[info] 
[info] Total time: 0 s, completed Nov 24, 2010 8:54:45 AM
```

Anyway, I think Cargo is very outstanding from all of those, as it's clear and informative, and the purescript seems not bad too.

**Comment by andylokandy** [comment]
Candidate: Combination of Cargo and pulp:

```
   Compiling Type.Data.RowList (idris-dev/contrib v1.0.2)
   Compiling Type.Data.Row (idris-dev/contrib v1.0.2)
   Compiling Data.Symbol (idris-dev/contrib v1.0.2)
   Compiling Data.Boolean (idris-dev/contrib v1.0.2)
   Compiling Control.Semigroupoid (andy/extra v0.1.0)
   Compiling Control.Category (andy/extra v0.1.0)
```

**Comment by dcao** [comment]
One option would just to keep what we have; the CLI output that we have now is growing on me a little...

Here's another candidate to consider: combining parts of Cargo/pulp with what we have right now:

```
$ elba test

     [1/5] Updating indices
Retrieving index+git+https://github.com/elba/index
    [warn] Some example warning
    Stored indices at ~/.elba/indices

     [2/5] Resolving dependencies
Retrieving idris-dev/contrib@dir+../contrib|1.0.2
Retrieving andy/extra@git+https://github.com/andylokandy/idris-extra|0.1.0
    Locked package dependency graph at elba.lock

     [3/5] Retrieving packages
Retrieving dcao/index-package@index+git+https://github.com/elba/index|0.1.2
           ...

     [4/5] Building targets
  Building idris-dev/contrib@dir+../contrib|1.0.2
           - Type.Data.RowList
           - Type.Data.Row
           - Data.Symbol

     [5/5] Running tests
   Running target/bin/test1
   Running target/bin/test2

    [info] Test results
    Passed test1
[stdout]
...
[stderr]
...
    Failed test2
[stdout]
...
[stderr]
...

error! Tests did not pass.

$ _
```

**Comment by andylokandy** [comment]
I think the first three steps is too verbose, like the sbt one. But I think we can still have it as a verbose version output. There are a few downsides of it:

- Usually, user only cares the parts that are slow or went wrong, the `Retrieving` , `Compiling` and `stderr`, to be specific. The other logs like where we store index or when we generate the lockfile happens so fast that the user has no time to read it and usually the user would not be interested in them, but they are actually distracting them from useful information, like test result.

- It doesn't help making elba more easy to use. Because in order to understanding what the `Locked package dependency graph at elba.lock` means, user have to learn how elba resolve packages in advance. Also, the `Stored indices at ~/.elba/indices` may confuse the users if they don't know what `indices` is. In fact, the less concepts user has to understand, the more pleasant for user to work with. For example, you are rarely to know where Cargo stores the index in your first month with rust.

**Comment by dcao** [comment]
Alright then, in the next commit I'll use the output method you suggested (and create a new utility struct for handling printing)

**Comment by dcao** [comment]
Resolved by [ca24df95...](https://github.com/elba/elba/commit/ca24df95ffcffbe8519bc5193bd3123593eb3bfb)
